### PR TITLE
Feature/token authentication

### PR DESCRIPTION
**Pull request by MartinMeyer1**

### CODE REVIEW

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 1]
```diff
@@ -190,6 +199,31 @@
  *         the very same way as it is done by HTTPS. If "false" the host name is not verified at all. Per default the
  *         hostname is verified (true).
  *     </li>
+ *
+ *     <li>
+ *         <b>ch.hevs.cloudio.endpoint.token</b><br>
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:204]: It is for provisioning endpoints, so I. would prefer the key `ch.hevs.cloudio.endpoint.provisioningToken`

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 2]
```diff
@@ -190,6 +199,31 @@
...
+ *
+ *         This property contains a single-use authentication token for the endpoint. This token is used to retrieve
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:206]: This property contains a single-use authentication token for endpoint provisioning. This token is used to retrieve certificates and settings using the cloud.iO Rest API.

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 3]
```diff
@@ -190,6 +199,31 @@
...
+ *         <b>ch.hevs.cloudio.endpoint.tokenPropertiesPath</b><br>
+ *
+ *         This property is used to specify a path where the retrieved certificates and .properties file from token
+ *         authentication will be stored. <br>
+ *
+ *         This property is optional and taken into account only if token authentication is used.
+ *         Its the default is <b>/etc/cloud.io/</b>.
+ *     </li>
+ *     <li>
+ *         <b>ch.hevs.cloudio.endpoint.apiUrl</b><br>
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:220]: `ch.hevs.cloudio.endpoint.provisioningUrl`

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 4]
```diff
@@ -190,6 +199,31 @@
...
+ *
+ *          Main URL of the cloud.iO Rest API. This URL is used for the certificate retrieval using authentication
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:222]: Main URL of the cloud.iO Rest API. This URL is used for the certificate and settings retrieval using one-time provisioning with using a provisioning token.<br>

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 5]
```diff
@@ -190,6 +199,31 @@
...
+ *
+ *     <li>
+ *         <b>ch.hevs.cloudio.endpoint.token</b><br>
+ *
+ *         This property contains a single-use authentication token for the endpoint. This token is used to retrieve
+ *         certificates using the cloud.iO Rest API.
+ *
+ *     </li>
+ *     <li>
+ *         <b>ch.hevs.cloudio.endpoint.tokenPropertiesPath</b><br>
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:211]: `ch.hevs.cloudio.endpoint.provisioningSavePath`

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 6]
```diff
@@ -190,6 +199,31 @@
...
+ *
+ *         This property is used to specify a path where the retrieved certificates and .properties file from token
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:213]: This property is used to specify a path where the retrieved certificates and .properties file from the provisioning process will be saved.<br>

This property is optional and taken into account only if token based provisioning is used.
The default value is <b>~/.config/cloud.io/</b>.

IMHO `~/.config/cloud.io/` is a better default as we are not sure the the endpoint is running as root.

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 7]
```diff
@@ -465,6 +499,12 @@ class InternalEndpoint implements CloudioNodeContainer, MqttCallback, Runnable,
         private static final String PERSISTENCE_MQTT_UPDATE         = "cloudioPersistenceUpdate";
         private static final String PERSISTENCE_MQTT_LOG            = "cloudioPersistenceLog";
         private static final String PERSISTENCE_MQTT_LIFECYCLE      = "cloudioPersistenceLifecycle";
+        /*** TOKEN parameters******************************************************************************************/
+        private static final String AUTH_TOKEN = "ch.hevs.cloudio.endpoint.token";
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:503]: `private` static final String PROVISIONING_TOKEN_PROPERTY     = "ch.hevs.cloudio.endpoint.provisioningToken";`

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 8]
```diff
@@ -465,6 +499,12 @@ class InternalEndpoint implements CloudioNodeContainer, MqttCallback, Runnable,
...
+        private static final String AUTH_TOKEN_PROPERTIES_PATH = "ch.hevs.cloudio.endpoint.tokenPropertiesPath";
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:504]: `private static final String PROVISIONING_SAVE_PATH_PROPERTY = "ch.hevs.cloudio.endpoint.provisioningSavePath";
        `

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 9]
```diff
@@ -465,6 +499,12 @@ class InternalEndpoint implements CloudioNodeContainer, MqttCallback, Runnable,
...
+        private static final String TOKEN_PROPERTIES_DEFAULT_PATH = "/etc/cloud.io/";
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:505]: `private static final String TOKEN_PROPERTIES_DEFAULT_PATH = "~/.config/cloud.io/";`

Default path should work even when not  running  as root.

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 10]
```diff
@@ -465,6 +499,12 @@ class InternalEndpoint implements CloudioNodeContainer, MqttCallback, Runnable,
...
+        private static final String TOKEN_PROPERTIES_FILENAME = "token.properties";
+        private static final String API_URL = "ch.hevs.cloudio.endpoint.apiUrl";
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:507]: `private static final String PROVISIONING_URL_PROPERTY       = "ch.hevs.cloudio.endpoint.provisioningUrl";`

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 11]
```diff
@@ -523,6 +563,138 @@ public InternalEndpoint(String uuidOrAppName, CloudioEndpointConfiguration confi
...
+                    try {
+
+                        String mainURL = configuration.getProperty(API_URL, null);
+
+                        if (mainURL == null) {
+                            throw new InvalidPropertyException("CloudioEndpoint properties missing: No " + API_URL +
+                                    " properties given as argument");
+                        }
+
+                        HttpRequest request = HttpRequest.newBuilder()
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:589]: `HttpRequest request = HttpRequest.newBuilder()
                                .uri(new URI(mainURL + "/api/v1/provision/" + token))
                                .GET()
                                .header("Accept", "application/java-archive")
                                .timeout(Duration.of(10, ChronoUnit.SECONDS))
                                .build();`

New API will use the Accept header  to select the format.

**Review by cm0x4D — CHANGES_REQUESTED** [review]
Ideally the provisioning should provide all settings for the endpoint.

We should have a quick discussion how such a provisioning will take place and then finish to implement server and endpoint side accordingly.

**Comment by MartinMeyer1** [comment]
@cm0x4D The provisioning can already contains all the properties if you want to. You can add custom properties when creating a token:
image

**Comment by cm0x4D** [comment]
Yes, it is possible.  It is still not very user friendly. Most of the parameters should be automatically filled in by the server itself.  In the mean time we can do everything by hand. As the provisioning is a complete separate process from the normal operation, I propose that we add a main function to the java endpoint lib that does nothing but the provisioning. Then the endpoint can be started with that configuration. What do you think about that, @MartinMeyer1 ?

**Comment by MartinMeyer1** [comment]
For now, the server fills "ch.hevs.cloudio.endpoint.ssl.authorityPassword" and "ch.hevs.cloudio.endpoint.ssl.clientPassword". I think we can add "ch.hevs.cloudio.endpoint.uuid" and "ch.hevs.cloudio.endpoint.hostUri" (or maybe not if the server is behind a proxy?) by default. But don't you think that the other properties are more "project specific" and should not be filled automatically by the server?
To be sure to understand, your proposal is to have a provisioning class with its own ".properties" file, that provides the endpoint's ".properties"?
I think that having 2 separate files is cleaner than having 2 ".properties" that are then merged.

**Comment by cm0x4D** [comment]
It would be good to fill in the defaults for important settings by the server and let the user override them. But you are right, there are not many others.

No it is rather a program that receives the settings and uses only program arguments, no properties files at all. I am working on a proof of concept...

**Comment by lucblender** [comment]
new pr for token provisioning here #58

**Pull request closed without merging**